[AArch64][GlobalISel] Basic add_sat and sub_sat vector handling.

davemgreen opened the pull request:
This tries to fill in the basic vector handling for sadd_sat/uadd_sat and ssub_sat/usub_sat. It just handles the basics, marking legal types and clamping illegally sized vectors to legal ones.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1153,8 +1150,14 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
+      .clampNumElements(0, v2s32, v4s32)
+      .clampMaxNumElements(0, s64, 2)
+      .moreElementsToNextPow2(0)
+      .lowerIf(isScalar(0));
```
just make an unconditional .lower()?

davemgreen replied:
I was attempting to focus on the more normal vector types in this patch. A lot of other operations like G_SADDO do not handle vectors yet, let alone odd sized vectors, so it would seem difficult to write a test that works.

arsenm replied:
You can't really go wrong with a default case of lower

davemgreen replied:
I personally don't think we should be adding things if we don't have test coverage for them (or they don't work, like here). I was hoping this patch could handle the more standard types without trying to solve every problem.

arsenm replied:
It's more like having a trailing else if that can just be an else

davemgreen replied:
Yeah - it is a 'else' instead of 'unreachable'. I would just personally prefer that this failed during legalization (and fell back to SDAG) than it being untested, or tried to handle every different case in a single patch. I can change it to a lower though, we should get test coverage when the other operations are able to handle non-power-2 vector sizes.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1153,8 +1150,14 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
-  getActionDefinitionsBuilder({G_SADDSAT, G_SSUBSAT}).lowerIf(isScalar(0));
+  getActionDefinitionsBuilder({G_UADDSAT, G_SADDSAT, G_USUBSAT, G_SSUBSAT})
+      .legalFor({v2s64, v2s32, v4s32, v4s16, v8s16, v8s8, v16s8})
+      .clampNumElements(0, v8s8, v16s8)
```
` .widenScalarOrEltToNextPow2(0)`

tschuett replied:
Up to you, but the suggestion was about weird integers ala i6.

aemerson replied:
I'm not sure if it makes a difference between doing that and lowering, since after lowering the weird types will be handled again.

aemerson commented on llvm/test/CodeGen/AArch64/uadd_sat_vec.ll:
```diff
@@ -2,28 +2,11 @@
-; CHECK-GI-NEXT:  warning: Instruction selection used fallback path for v32i16
-; CHECK-GI-NEXT:  warning: Instruction selection used fallback path for v8i8
-; CHECK-GI-NEXT:  warning: Instruction selection used fallback path for v4i8
+; CHECK-GI:       warning: Instruction selection used fallback path for v4i8
```
I thought this would be handled by the elements clamp?

davemgreen replied:
I believe this is the G_ANYEXT issue we have seen elsewhere.

davemgreen committed:
[AArch64][GlobalISel] Basic add_sat and sub_sat vector handling.

This tries to fill in the basic vector handling for sadd_sat/uadd_sat and
ssub_sat/usub_sat. It just handles the basics, marking legal types and clamping
illegally sized vectors to legal ones.

davemgreen committed:
Use lower instead of lowerIf(scalar)

davemgreen commented:
I've rebased over the recent anyext fixes. Thanks

arsenm commented on llvm/test/CodeGen/AArch64/uadd_sat_vec.ll:
```diff
@@ -135,16 +145,39 @@ define void @v8i8(ptr %px, ptr %py, ptr %pz) nounwind {
+; CHECK-SD-NEXT:    str s0, [x2]
+; CHECK-SD-NEXT:    ret
+;
+; CHECK-GI-LABEL: v4i8:
```
I assume further patches will catch up to the DAG output

davemgreen replied:
Yeah that's the plan. It should hopefully be better in the end because we use widening, not expand.

davemgreen commented:
Thanks
